src/JsonApi/Serializer/ItemNormalizer.php

```diff
@@ -0,0 +1,561 @@
+            return;
+        }
+
+        if (!isset($data['data'])) {
+            throw new InvalidArgumentException(
+                'Key \'data\' expected. Only resource linkage currently supported, see: http://jsonapi.org/format/#document-resource-object-linkage'
+            );
+        }
+
+        $data = $data['data'];
+
+        if (!is_array($data) || 2 !== count($data)) {
+            throw new InvalidArgumentException(
+                'Only resource linkage supported currently supported, see: http://jsonapi.org/format/#document-resource-object-linkage'
+            );
+        }
+
+        if (!isset($data['id'])) {
+            throw new InvalidArgumentException(
+                'Only resource linkage supported currently supported, see: http://jsonapi.org/format/#document-resource-object-linkage'
+            );
+        }
+
+        return $this->itemDataProvider->getItem(
+            $this->resourceClassResolver->getResourceClass(null, $className),
+            $data['id']
+        );
+    }
+
+    /**
+     * Normalizes a relation as resource linkage relation.
+     *
+     * See: http://jsonapi.org/format/#document-resource-object-linkage
+     *
+     * For example, it may return the following array:
+     *
+     * [
+     *     'data' => [
+     *         'type' => 'dummy',
+     *         'id' => '1'
+     *     ]
+     * ]
+     *
+     * @param PropertyMetadata $propertyMetadata
+     * @param mixed            $relatedObject
+     * @param string           $resourceClass
+     * @param string|null      $format
+     * @param array            $context
+     *
+     * @return string|array
+     */
+    protected function normalizeRelation(
+        PropertyMetadata $propertyMetadata,
+        $relatedObject,
+        string $resourceClass,
+        string $format = null,
+        array $context
+    ) {
+        $resourceClass = $this->resourceClassResolver->getResourceClass(
+            $relatedObject,
+            null,
+            true
+        );
+
+        $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
+
+        $identifier = $this->getIdentifierFromItem($relatedObject);
+
+        return ['data' => [
+            'type' => $resourceMetadata->getShortName(),
+            'id' => (string) $identifier,
+        ]];
+    }
+
+    private function getIdentifierFromItem($item)
+    {
+        $identifiers = $this->getIdentifiersFromItem($item);
+
+        if (count($identifiers) > 1) {
+            throw new RuntimeException(sprintf(
+                'Multiple identifiers are not supported during serialization of relationships (Entity: \'%s\')',
+                $resourceClass
+            ));
+        }
+
+        return reset($identifiers);
+    }
+
+    /**
+     * Find identifiers from an Item (Object).
+     *
+     * Taken from ApiPlatform\Core\Bridge\Symfony\Routing\IriConverter
+     *
+     * TODO: Review if this would be useful if defined somewhere else
```

Comment:
We extracted this to a new file in this PR https://github.com/api-platform/core/pull/997/files You'll be able to decorate it when it get's merge. For now, let's keep this here.